MinGL.cpp

```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 	glDisable(GL_SCISSOR_TEST);
 }
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
+	int pixelSize = 4; // RGBA has 4 components (R, G, B, A)
+    unsigned char* pixel = new unsigned char[pixelSize];
+    glReadPixels(x, y, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, pixel);
+	const MinGLColor color{(float)pixel[0]/255, (float)pixel[1]/255, (float)pixel[2]/255, 1.0f};
+	return color;
+}
+
+void MinGL::drawLine(int x0, int y0, int x1, int y1, const MinGLColor& color){
+	// Bresenham's line generation algorithm is used here
+
+	bool steep = false;
+	if (std::abs(x0 - x1) < std::abs(y0 - y1)) {
+		std::swap(x0, y0);
+		std::swap(x1, y1);
+		steep = true;
+	}
+	if (x0 > x1) {
+		std::swap(x0, x1);
+		std::swap(y0, y1);
+	}
+	int dx = x1 - x0;
+	int dy = y1 - y0;
+	int derror = std::abs(dy) * 2;
+	int error = 0;
+	int y = y0;
+	for (int x = x0; x <= x1; ++x) {
+		if (steep) {
+			putPixel(y, x, color);
+		}
+		else {
+			putPixel(x, y, color);
+		}
+		error += derror;
+		if (error > dx) {
+			y += (y1 > y0 ? 1 : -1);
+			error -= dx * 2;
+		}
+	}
+}
+
+void MinGL::drawRectangle(int x0, int y0, int x1, int y1, const MinGLColor& color){
+	// specify any two diagonally opposite points of the rectangle
+
+	int xa = std::min(x0, x1);
+	int ya = std::min(y0, y1);
+
+	int xb = std::max(x0, x1);
+	int yb = std::max(y0, y1);
+
+	for(int i=xa; i<xb; i++){
+		putPixel(i, ya, color);
+	}
+	for(int i=ya; i<yb; i++){
+		putPixel(xb, i, color);
+	}
+	for(int i=xb; i>xa; i--){
+		putPixel(i, yb, color);
+	}
+	for(int i=yb; i>ya; i--){
+		putPixel(xa, i, color);
+	}
+}
+
+
+void MinGL::drawCircle(int xc, int yc, int rad, const MinGLColor& color){
+	// specify center (x,y) of circle and its radius
+	// Bresenham's circle drawing algorithm is used here
+
+	auto draw = [&](int x, int y){
```

Comment:
No need to capture everything by reference here, since most are just integers. If you must, capture only color by reference.